llvm/lib/Target/AIE/AIEPostPipeliner.cpp

```diff
@@ -188,54 +189,141 @@ int PostPipeliner::fit(MachineInstr *MI, int First, int Last, int II) {
   return -1;
 }
 
-void PostPipeliner::computeLoopCarriedParameters() {
-  // We schedule the first iteration, only using earliest. This updates
-  // earliest of the successors. Any successor in the second iteration
-  // represents a loop carried dependence, and we account for that by
-  // propagating its Earliest back to the first iteration
-  // Note that we don't have to clean the effects of this exploration,
-  // since the real scheduling will overwrite Cycle, and the ultimate Earliest
-  // will never be less than we compute here.
-
+void PostPipeliner::computeForward() {
+  // The forward order defines a topological sort, so we can compute
+  // Earliest and Ancestors in a single forward sweep
+  // Also initialize Latest to the last cycle in the last stage
+  const int Latest = -1;
   for (int K = 0; K < NInstr; K++) {
```

Comment:
Nit: I'd find all those traversals much easier to read if we didn't have to use array indexing.
E.g. something like
```
for (NodeInfo &Node : Info) {
  ...
  for (NodeInfo &Pred : Node.preds()) { ... }
}
```